src/csharp/Grpc.Core/Internal/UserAgentStringProvider.cs

```diff
@@ -0,0 +1,114 @@
+#region Copyright notice and license
+
+// Copyright 2021 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System.Collections.Generic;
+using System.Text.RegularExpressions;
+
+namespace Grpc.Core.Internal
+{
+    /// <summary>
+    /// Helps constructing the grpc-csharp component of the user agent string.
+    /// </summary>
+    internal class UserAgentStringProvider
+    {
+        static readonly UserAgentStringProvider defaultInstance;
+        readonly string userAgentString;
+
+        static UserAgentStringProvider()
+        {
+            defaultInstance = new UserAgentStringProvider(VersionInfo.CurrentVersion, PlatformApis.FrameworkDescription, PlatformApis.ClrVersion, PlatformApis.GetGrpcCoreTargetFrameworkMoniker(), PlatformApis.ProcessArchitecture);
+        }
+
+        public static UserAgentStringProvider DefaultInstance => defaultInstance;
+
+        public string GrpcCsharpUserAgentString => userAgentString;
+
+        public UserAgentStringProvider(string grpcCsharpVersion, string frameworkDescription, string clrVersion, string tfm, CommonPlatformDetection.CpuArchitecture arch)
+        {
+            var detailComponents = new List<string>();
+
+            string sanitizedFrameworkDescription = SanitizeFrameworkDescription(frameworkDescription);
+            if (sanitizedFrameworkDescription != null)
+            {
+                detailComponents.Add(sanitizedFrameworkDescription);
+            }
+
+            if (clrVersion != null)
+            {
+                detailComponents.Add($"CLR {clrVersion}");
+            }
+
+            if (tfm != null)
+            {
+                detailComponents.Add(tfm);
+            }
+
+            string architectureString = TryGetArchitectureString(arch);
+            if (architectureString != null)
+            {
+                detailComponents.Add(architectureString);
+            }
+
+            // TODO(jtattermusch): consider adding details about running under unity / xamarin etc.
+            var details = string.Join("; ", detailComponents);
+            userAgentString = $"grpc-csharp/{grpcCsharpVersion} ({details})";
+        }
+
+        static string TryGetArchitectureString(CommonPlatformDetection.CpuArchitecture arch)
+        {
+            switch (arch)
```

Comment:
nit: can we get rid of this switch statement and just call `ToString()` on these enums? Then, we also don't need to check if this is null when constructing the user agent (I believe the architecture will just say "Unknown", which may be useful anyways)